bokehjs/src/lib/models/widgets/tables/cell_formatters.ts

```diff
@@ -81,6 +81,63 @@ export class StringFormatter extends CellFormatter {
   }
 }
 
+export namespace ScientificFormatter {
+  export type Attrs = p.AttrsOf<Props>
+
+  export type Props = StringFormatter.Props & {
+    precision: p.Property<number | 10>
+    power_limit_high: p.Property<number>
+    power_limit_low: p.Property<number>
+  }
+}
+
+export interface ScientificFormatter extends ScientificFormatter.Attrs {}
+
+export abstract class ScientificFormatter extends StringFormatter {
+  properties: ScientificFormatter.Props
+
+  constructor(attrs?: Partial<ScientificFormatter.Attrs>) {
+    super(attrs)
+  }
+
+  static init_ScientificFormatter(): void {
+    this.define<ScientificFormatter.Props>({
+      precision:        [ p.Any,     10     ],
+      power_limit_high: [ p.Number,  5      ],
+      power_limit_low:  [ p.Number,  -3     ],
+    })
+  }
+
+  get scientific_limit_low(): number {
+    return Math.pow(10.0, this.power_limit_low)
+  }
+
+  get scientific_limit_high(): number {
+    return Math.pow(10.0, this.power_limit_high)
+  }
+
+  doFormat(row: any, cell: any, value: any, columnDef: any, dataContext: any): string {
+    let need_sci = value >= this.scientific_limit_high || value <= this.scientific_limit_low
+    let precision = this.precision
+
+    if (precision == null || isNumber(precision)) {
```

Comment:
`this.precision` is of type `number` (barring the ` | 10`, which would still give `number`), so neither the `null` check, nor `isNumber()` check are necessary. TypeScript doesn't complain about both, because `null` checks are generally always allowed (due to reasons, which I don't necessarily agree with), but due to strict null checks, `null` won't be ever allowed to be assigned to `this.precision`, and `isNumber(precision)` is a no-op, yet a valid check nevertheless.

Reply:
If you have a concern that something wouldn't work at runtime, then we handle all such issues at validation time in our properties system. Then our code can be free of any defensive and unnecessary checks.

Reply:
However, if it is desired to allow precision to be nullable (optional), then this needs to be ended in the the type of the property, i.e. `p.Property<number | null>`. We use strict null checks, so each `null` or `undefined` must be explicitly mentioned in the type of an entity.

Reply:
ok, thanks for the explanation. I understand the typescript code a bit better now and see why that is not needed here, but is needed in the BasicTickFormatter, which also has the auto option